TYPO3.Neos/Classes/TYPO3/Neos/Utility/NodeUriPathSegmentGenerator.php

```diff
@@ -31,22 +33,44 @@ class NodeUriPathSegmentGenerator
     protected $transliterationService;
 
     /**
+     * @Flow\Inject
+     * @var NodeTypeManager
+     */
+    protected $nodeTypeManager;
+
+    /**
+     * @Flow\Inject
+     * @var NodeSearchServiceInterface
+     */
+    protected $nodeSearchService;
+
+    /**
      * Sets the best possible uriPathSegment for the given Node.
      * Will use an already set uriPathSegment or alternatively the node name as base,
      * then checks if the uriPathSegment already exists on the same level and appends a counter until a unique path segment was found.
      *
      * @param NodeInterface $node
      * @return void
      */
-    public static function setUniqueUriPathSegment(NodeInterface $node)
+    public function setUniqueUriPathSegment(NodeInterface $node)
     {
         if ($node->getNodeType()->isOfType('TYPO3.Neos:Document')) {
-            $q = new FlowQuery(array($node));
-            $q = $q->context(array('invisibleContentShown' => true, 'removedContentShown' => true, 'inaccessibleContentShown' => true));
-
+            $parentNode = $node->getParent();
+            $documentNodeType = $this->nodeTypeManager->getNodeType('TYPO3.Neos:Document');
+            $context = $node->getContext();
             $possibleUriPathSegment = $initialUriPathSegment = !$node->hasProperty('uriPathSegment') ? $node->getName() : $node->getProperty('uriPathSegment');
+
+            $nodeExists = function ($possibleUriPathSegment) use ($documentNodeType, $context, $parentNode, $node) {
+                $nodes = $this->nodeSearchService->findByProperties(['uriPathSegment' => $possibleUriPathSegment], [$documentNodeType], $context, $parentNode);
```

Comment:
Thinking about this one again I realize that it definitely can have a huge performance penalty. Since it actually does a like query on all documents nodes when trying to find those matching on the first level. So the further up and the bigger the tree the worse it gets. So at the very least, the query needs to be limited to a certain level to avoid that problem.

Regarding it being faster being faster just to loop, I agree it's probably little difference. However data is king, let's not make assumptions when it can easily be tested please.